applications/luci-app-shadowsocks-libev/htdocs/luci-static/resources/view/shadowsocks-libev/rules.js

```diff
@@ -0,0 +1,124 @@
+			_('On this page you can configure how traffics are to be \
+				forwarded to ss-redir instances. \
+				If enabled, packets will first have their src ip addresses checked \
+				against <em>Src ip/net bypass</em>, <em>Src ip/net forward</em>, \
+				<em>Src ip/net checkdst</em> and if none matches <em>Src default</em> \
+				will give the default action to be taken. \
+				If the prior check results in action <em>checkdst</em>, packets will continue \
+				to have their dst addresses checked.'));
+
+		s = m.section(form.NamedSection, 'ss_rules', 'ss_rules');
+		s.tab('general', _('General Settings'));
+		s.tab('src', _('Source Settings'));
+		s.tab('dst', _('Destination Settings'));
+
+		s.taboption('general', form.Flag, 'disabled', _('Disable'));
+		if (!stats[1]) {
+			ss.option_install_package(s, 'general');
+		}
+
+		o = s.taboption('general', form.ListValue, 'redir_tcp',
+			_('ss-redir for TCP'));
+		ss.values_redir(o, 'tcp');
+		o = s.taboption('general', form.ListValue, 'redir_udp',
+			_('ss-redir for UDP'));
+		ss.values_redir(o, 'udp');
+
+		o = s.taboption('general', form.ListValue, 'local_default',
+			_('Local-out default'),
+			_('Default action for locally generated TCP packets'));
+		ss.values_actions(o);
+		o = s.taboption('general', widgets.DeviceSelect, 'ifnames',
+			_('Ingress interfaces'),
+			_('Only apply rules on packets from these network interfaces'));
+		o.multiple = true;
+		o.noaliases = true;
+		s.taboption('general', form.Value, 'ipt_args',
+			_('Extra arguments'),
+			_('Passes additional arguments to iptables. Use with care!'));
+
+		src_dst_option(s, 'src', form.DynamicList, 'src_ips_bypass',
+			_('Src ip/net bypass'),
+			_('Bypass ss-redir for packets with src address in this list'));
+		src_dst_option(s, 'src', form.DynamicList, 'src_ips_forward',
+			_('Src ip/net forward'),
+			_('Forward through ss-redir for packets with src address in this list'));
+		src_dst_option(s, 'src', form.DynamicList, 'src_ips_checkdst',
+			_('Src ip/net checkdst'),
+			_('Continue to have dst address checked for packets with src address in this list'));
+		o = s.taboption('src', form.ListValue, 'src_default',
+			_('Src default'),
+			_('Default action for packets whose src address do not match any of the src ip/net list'));
+		ss.values_actions(o);
+
+		src_dst_option(s, 'dst', form.DynamicList, 'dst_ips_bypass',
+			_('Dst ip/net bypass'),
+			_('Bypass ss-redir for packets with dst address in this list'));
+		src_dst_option(s, 'dst', form.DynamicList, 'dst_ips_forward',
+			_('Dst ip/net forward'),
+			_('Forward through ss-redir for packets with dst address in this list'));
+
+		var dir = '/etc/shadowsocks-libev';
+		o = s.taboption('dst', form.FileUpload, 'dst_ips_bypass_file',
+			_('Dst ip/net bypass file'),
+			_('File containing ip/net for the purposes as with <em>Dst ip/net bypass</em>'));
+		o.datatype = 'file';
+		o.root_directory = dir;
+		s.taboption('dst', form.FileUpload, 'dst_ips_forward_file',
```

Comment:
Missing assignment to `o`.  Also the datatype spec is likely implied by FileUpload widget?